packages/blocks/src/api/types.ts

```diff
@@ -0,0 +1,27 @@
+/**
+ * WordPress dependencies
+ */
+import type { WPElement } from '@wordpress/element';
+
+export type BlockElementNode = {
+	type: WPElement[ 'type' ];
+	props: {
+		attributes: Record< string, any >;
+		children: BlockChildren;
+	};
+};
+
+/**
+ * A representation of a single node within a block's rich text value. If
+ * representing a text node, the value is simply a string of the node value.
+ * As representing an element node, it is an object of:
+ *
+ * 1. `type` (string): Tag name.
+ * 2. `props` (object): Attributes and children array of BlockNode.
+ */
+export type BlockNode = BlockElementNode | string;
```

Comment:
note that we talk about a "BlockNode" as being the result of the initial parse of a Gutenberg serialized post. this is a block type, set of attributes, inner blocks, and inner HTML.

is that what we're referring to here? It seems like we're talking about something more HTML-specific…

Reply:
I used the type name used but not defined in [node.js](https://github.com/WordPress/gutenberg/blob/bd12cc78183489bdf2db036ececf8e96bbe4d637/packages/blocks/src/api/node.js#L26) and [children.js](https://github.com/WordPress/gutenberg/blob/bd12cc78183489bdf2db036ececf8e96bbe4d637/packages/blocks/src/api/children.js#L40). Though it looks like those are depreciated in favor `html` in [machers.js](https://github.com/WordPress/gutenberg/blob/bd12cc78183489bdf2db036ececf8e96bbe4d637/packages/blocks/src/api/matchers.js#L12).

Would could help inform this process is a very well defined lifecycle of a block, and some consisense on naming.

Reply:
not sure. this is `RichText` stuff isn't it?

`BlockDOMNode`?
Something like that?

Reply:
cc: @ellatrix @sirreal any thoughts? what _is_ this?

Reply:
@dmsnell after look at the declarations in [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/b9748855c21f65681c228df1fb60e72ce489c1d3/types/wordpress__blocks/api/node.d.ts#L36), the parameter `node` (`WPBlockNode` in original JSDocs) is typed as `JSX.Element` on DT. Should the `Object` in the JSDoc below  be `JSX.Element`? Or possibly more appropriately `WPElement`?

https://github.com/WordPress/gutenberg/blob/bd12cc78183489bdf2db036ececf8e96bbe4d637/packages/blocks/src/api/node.js#L11-L20

Reply:
@dmsnell I just found this while reviewing unresolved comments. I matches the types from DT. Take a look at it and see if you think it is correct. The types on DT did not declare a `RichTextElement` type, but typed it as a `JSX.Element`, I will defer to you on what is correct.

Reply:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/wordpress__blocks/api/node.d.ts

Reply:
@dmsnell I'm sorry I missed this comment, I was reviewing all the unresolved comments and found this. I copied the types from [@types/wordpress__blocks](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/wordpress__blocks/api/node.d.ts). Though They do note match with the original `WPBlockNode` type of the JSDoc